### DISCUSSION TITLE
How to scroll in Android with 2 layout in Tablet Devices

### DISCUSSION THREAD

**Discussion by yukender** [appium/appium#19370] [Q&A]
Hi, 

We are automating cases in Mobile and we are testing both in Landscape and Portrait modes.

In portrait  mode we have two layout and in landscape it will be three layout.

We use MobileBy.AndroidUIAutomator( "new UiScrollable(new UiSelector()" + ".scrollable(true)).scrollIntoView("+ "new UiSelector().text(\"" + expectedText + "\"));")  method to scroll the elements but we cannot scroll in Android tablets with two and three layout mode.

Appium Inspector framelayout :-
![image](https://github.com/appium/appium/assets/62298454/dc108cfe-35dc-46cb-8605-7a091c568e86)

Java Version :- 8
Appium Version :- 1.22.2

Sample Image :- 
![image](https://github.com/appium/appium/assets/62298454/8dd844e4-1fd6-4b64-b519-7f23fe46eaee)

Does any solution can be provided like does i need to upgrade java or appium, any new scroll can be used to differentiate the layout by framelayout?

**Comment by portyu9** [comment]
Hi @yukender 

The Java version is probably not the reason for this behavior.

The important part of your locator is this:

```java
new UiSelector().scrollable(true)
```

On a phone with only one scrollable area this often works because there is only one match. On a tablet with two or three panes, there can be **multiple scrollable containers**, so the generic selector may scroll the wrong one.

Instead, identify the actual scrollable container for the pane you want to use and make the `UiScrollable` selector specific.

For example, if the pane has a resource ID:

```java
driver.findElement(AppiumBy.androidUIAutomator(
    "new UiScrollable(new UiSelector()" +
    ".resourceId(\"com.example:id/right_list\")" +
    ".scrollable(true))" +
    ".scrollIntoView(new UiSelector().text(\"" + expectedText + "\"))"
));
```

If the individual scrollable panes have no unique resource ID, `instance()` can distinguish them:

```java
driver.findElement(AppiumBy.androidUIAutomator(
    "new UiScrollable(new UiSelector()" +
    ".scrollable(true).instance(1))" +
    ".scrollIntoView(new UiSelector().text(\"" + expectedText + "\"))"
));
```

However, `instance(1)` / `instance(2)` is more fragile because the ordering can change between portrait and landscape. A unique `resource-id` or accessibility identifier on each pane is preferable.

Also, a `FrameLayout` itself is not necessarily the object that should be scrolled. Inspect the hierarchy and find the actual scrollable child (`RecyclerView`, `ScrollView`, etc.) that has `scrollable=true`.

There was also an Appium/UiAutomator2 issue in older Appium 1.x versions where scrolling in screens containing multiple scrollable views could target the first available view instead of the intended one. Since you are using Appium `1.22.2`, upgrading to a current Appium/UiAutomator2 setup is strongly recommended.

With current UiAutomator2 versions you can also target a particular scrollable element with the supported mobile scrolling APIs instead of relying on an unqualified `scrollable(true)` selector.

So I would do these in order:

1. Give each tablet pane's scrollable view a unique `resource-id` if you control the app.
2. Target that scrollable view explicitly in `UiScrollable`.
3. Use `instance()` only as a fallback.
4. Upgrade from Appium 1.x rather than changing Java purely to solve this scrolling issue.

Please mark this answer as accepted if it helped, thank you.
